packages/back-end/src/services/clickhouse.ts

```diff
@@ -0,0 +1,95 @@
+import * as crypto from "crypto";
+import { createClient as createClickhouseClient } from "@clickhouse/client";
+import {
+  CLICKHOUSE_HOST,
+  CLICKHOUSE_ADMIN_USER,
+  CLICKHOUSE_ADMIN_PASSWORD,
+  CLICKHOUSE_DATABASE,
+} from "back-end/src/util/secrets";
+import { DataSourceParams } from "back-end/types/datasource";
+import { ReqContext } from "back-end/types/organization";
+import { logger } from "back-end/src/util/logger";
+
+function generatePassword(length: number = 16): string {
+  const uppercaseChars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+  const specialChars = "!@#$%^&*()_+[]{}|;:,.<>?";
+  const numericChars = "0123456789";
+  const allChars =
+    "abcdefghijklmnopqrstuvwxyz" + uppercaseChars + numericChars + specialChars;
+
+  let password = "";
+  password += uppercaseChars[Math.floor(Math.random() * uppercaseChars.length)];
+  password += specialChars[Math.floor(Math.random() * specialChars.length)];
+  password += numericChars[Math.floor(Math.random() * numericChars.length)];
+
+  for (let i = 3; i < length; i++) {
+    password += allChars[Math.floor(Math.random() * allChars.length)];
+  }
+
+  return password
+    .split("")
+    .sort(() => 0.5 - Math.random())
+    .join("");
+}
```

Comment:
We should consider using a third party here like this: https://www.npmjs.com/package/generate-password. This package uses the node crypto module to generate random bytes.

Reply:
We should consider using a third party here like this: https://www.npmjs.com/package/generate-password. This package uses the node crypto module to generate random bytes.